### PR TITLE
HHH-12738 - Session/EntityManager is closed in ForeignGenerator (JTA setup)

### PR DESCRIPTION
**Pull request by mjesseni**
I just added the test case closing the entity manager, before the commit phase is stared. Personally, I would never close the entity manager by hand. This is just to simulate what Spring is doing when commiting a transaction. First it closes the resource (the entity manager) and afterwards performing the commit.

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/test/idgen/foreign/ForeignGeneratorResourceLocalTest.java [hunk 1]
```diff
@@ -36,275 +34,366 @@
...
+                Customer.class)
+                .setParameter("customerId", customerId)
+                .getSingleResult();
+
+            CustomerContractRelation relation = new CustomerContractRelation();
+            relation.setContractId(contractId);
+            customer.addContractRelation(relation);
+
+            // simulating spring JtaTransactionManager.triggerBeforeCompletion()
+            entityManager.close();
```
Comment (vladmihalcea) [inline hibernate-core/src/test/java/org/hibernate/test/idgen/foreign/ForeignGeneratorResourceLocalTest.java]: That's not a valid workflow. Do you want to span a Tx over multiple EntityManagers?

Comment (mjesseni) [reply hibernate-core/src/test/java/org/hibernate/test/idgen/foreign/ForeignGeneratorResourceLocalTest.java]: Hi Vlad! Thx for your quick reply. I am totally aware about that and I would not close the session as well. However, this is how Spring does the integration. 

Well, within the project I am working in, we have multiple entity managers and spawning Tx over them using Bitronix as XA transaction manager. This was the setup I originaly wanted to provid by my test case including the spring and bitronix dependencies. 

As a workaround for this issue I am able to manually flush the entity manager with the business code and everything is going well. Only if a left this work to the environment (spring), I run into this problems. However, the workaround seems to me not very practicable at all.

I am not sure how to proceed. As already mentioned, spring contributors are quite sure, that closing the entity manager before starting the commit phase should be no problem at all. And of cause, it seems to be no problem except the situation in the ForeignGenerator. 

Thx for your time and patience!

**Comment by vladmihalcea** [comment]
I rewrote your PR and could replicate the issue.

File: hibernate-core/src/test/java/org/hibernate/test/idgen/foreign/ForeignGeneratorResourceLocalTest.java (hunk 1 above)
Comment (vladmihalcea) [reply hibernate-core/src/test/java/org/hibernate/test/idgen/foreign/ForeignGeneratorResourceLocalTest.java]: @mjesseni I managed to replicate it, so I reopened it. I'll propose a fix today.

**Comment by vladmihalcea** [comment]
Replaced by [this PR](https://github.com/hibernate/hibernate-orm/pull/2377) which provides a fix as well.

File: hibernate-core/src/test/java/org/hibernate/test/idgen/foreign/ForeignGeneratorResourceLocalTest.java (hunk 1 above)
Comment (mjesseni) [reply hibernate-core/src/test/java/org/hibernate/test/idgen/foreign/ForeignGeneratorResourceLocalTest.java]: @vladmihalcea thx a lot!

**Pull request closed without merging**